src/util/side_metadata/helpers.rs

```diff
@@ -0,0 +1,210 @@
+
+#[cfg(target_pointer_width = "32")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 32;
+// FIXME: This must be updated if the heap layout changes
+#[cfg(target_pointer_width = "64")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 48;
+
+pub(super) const MAX_METADATA_BITS: usize = constants::BITS_IN_WORD;
+// const SPACE_PER_META_BIT: usize = 2 << (MAX_HEAP_SIZE_LOG - constants::LOG_BITS_IN_WORD);
+pub(super) const META_SPACE_PAGE_SIZE: usize = constants::BYTES_IN_PAGE;
+pub(super) const META_SPACE_PAGE_SIZE_LOG: usize = constants::LOG_BYTES_IN_PAGE as usize;
+
+#[inline(always)]
+pub(super) fn address_to_meta_address(addr: Address, metadata_id: SideMetadataID) -> Address {
+    let bits_num_log = METADATA_SINGLETON.meta_bits_num_log_vec[metadata_id.as_usize()] as i32;
+    // right shifts for `align` times, then
+    // if bits_num_log < 3, right shift a few more times to cover multi objects per metadata byte
+    // if bits_num_log = 3, metadata byte per object is 1
+    // for > 3, left shift, because more than 1 byte per object is required
+    let rshift = (constants::LOG_BITS_IN_BYTE as i32) - bits_num_log;
+    let offset = if rshift >= 0 {
+        addr.as_usize()
+            .checked_shr(METADATA_SINGLETON.align[metadata_id.as_usize()] as u32)
+            .unwrap()
+            .checked_shr(rshift as u32)
+            .unwrap()
+    } else {
+        addr.as_usize()
+            .checked_shr(METADATA_SINGLETON.align[metadata_id.as_usize()] as u32)
+            .unwrap()
+            .checked_shl(-rshift as u32)
+            .unwrap()
+    };
+
+    METADATA_SINGLETON.meta_base_addr_vec[metadata_id.as_usize()] + offset
+}
+
+// Gets the related meta address and clears the low order bits
+pub(super) fn address_to_meta_page_address(
+    data_addr: Address,
+    metadata_id: SideMetadataID,
+) -> Address {
+    let meta_addr = address_to_meta_address(data_addr, metadata_id);
+    unsafe {
+        Address::from_usize((meta_addr >> META_SPACE_PAGE_SIZE_LOG) << META_SPACE_PAGE_SIZE_LOG)
+    }
+}
+
+// Checks whether the meta page containing this address is already mapped.
+//
+// Returns Err if the address is not mappable by mmtk,
+// and Ok(is_mapped?) otherwise.
+//
+// NOTE: using incorrect (e.g. not properly aligned) page_addr is undefined behavior.
+pub(super) fn meta_page_is_mapped(page_addr: Address) -> Result<bool> {
+    let prot = libc::PROT_NONE;
+    // MAP_FIXED_NOREPLACE returns EEXIST if already mapped
+    let flags = libc::MAP_ANON | libc::MAP_PRIVATE | libc::MAP_FIXED_NOREPLACE;
+    let result: *mut libc::c_void = unsafe {
+        libc::mmap(
+            page_addr.to_mut_ptr(),
+            META_SPACE_PAGE_SIZE,
+            prot,
+            flags,
+            -1,
+            0,
+        )
+    };
+    if result == libc::MAP_FAILED {
+        let err = unsafe { *libc::__errno_location() };
+        if err == libc::EEXIST {
+            // mmtk already mapped it
```

Comment:
The virtual address space may be mapped by other threads/libs/ or a VM as long as they are in the same process as MMTK.
Afaik, mmap with the `MAP_FIXED_NOREPLACE ` flag can differentiate existing mappings from not-mappable areas.
We should not use the continuous area belonging to side metadata for any other purposes (by mmtk or vm). Considering this, holes in the area will only be caused by not-mappable addresses, which will return a failure code other than `EEXIST`.
Currently, such failures will only panic (It never happened in the 64-bit address space).
But ultimately, we need to change the whole memory allocation code, to consider the failures of metadata space mapping as it does for the data mapping (e.g. data mapping should be changed until metadata mapping succeeds).

Reply:
I still dont understand this.
1. Does the function try to make sure that the `addr` is used or to be used as metadata memory? If so, that means the memory should be either already mapped by mmtk, or not mapped at all. In other cases, we cannot use this `addr` as metadata memory.
2. Can `EEXIST` tell if this addr is mapped by mmtk or by VM? Because we need to differentiate that. If it is mapped by mmtk, that means we can use the memory for metadata. But if it is mapped by the VM, that means we cannot use the memory for metadata.

Reply:
`EEXIST` doesn't differentiate between mmtk and VM.
Side metadata needs an **EXCLUSIVE** address range for all its memory. This address range can have non-mappable holes, but it can't be mapped by any other threads in the mmtk's process, including the VM.
So, it must never happen that a part of the side metadata's address range is used by anything except the side metadata itself.